src/btree/bt_read.c

```diff
@@ -184,6 +184,300 @@ __bt_reconstruct_delta(WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *delta)
+
+    WT_ASSERT(session, i == base_entries && j == delta_entries);
+    WT_ASSERT(session, final_entries != 0);
+    WT_ASSERT(session, final_entries < estimated_entries && refs[final_entries] == NULL);
+
+    *ref_entriesp = final_entries;
+
+err:
+    __wt_free(session, base);
+    return (ret);
+}
+
+/*
+ * __bt_reconstruct_internal_deltas --
+ *     Reconstructs the internal page using `delta_size` delta images.
+ */
+static int
+__bt_reconstruct_internal_deltas(
+  WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *deltas, size_t delta_size)
+{
+    WT_CELL_UNPACK_DELTA_INT **unpacked_deltas, **unpacked_deltas_merged;
+    WT_DECL_RET;
+    WT_DELTA_HEADER *header;
+    WT_PAGE_INDEX *pindex;
+    WT_REF **refs;
+    size_t *delta_size_each, incr, pindex_size, refs_entries, unpacked_deltas_merged_size;
+    uint32_t i, j;
+
+    unpacked_deltas = unpacked_deltas_merged = NULL;
+    refs = NULL;
+    pindex = NULL;
+    unpacked_deltas_merged_size = refs_entries = incr = 0;
+
+    /*
+     * !!!
+     * Unpack all delta images into a 2D array where entry is WT_CELL_UNPACK_DELTA_INT:
+     *     unpacked_deltas[0] -> [ entry_0_0, entry_0_1, entry_0_2, ... ]
+     *     unpacked_deltas[1] -> [ entry_1_0, entry_1_1, entry_1_2, ... ]
+     *     unpacked_deltas[2] -> [ entry_2_0, entry_2_1, entry_2_2, ... ]
+     *                           ...
+     *     unpacked_deltas[N] -> [ entry_N_0, entry_N_1, entry_N_2, ... ]
+     */
+    WT_RET(__wt_calloc_def(session, delta_size, &delta_size_each));
+    WT_ERR(__wt_calloc_def(session, delta_size, &unpacked_deltas));
+    for (i = 0, j = 0; i < (uint32_t)delta_size; ++i, j = 0) {
+        header = (WT_DELTA_HEADER *)deltas[i].data;
+        WT_ASSERT(session, header->u.entries != 0);
+        delta_size_each[i] = (size_t)header->u.entries;
+
+        WT_ERR(__wt_calloc_def(session, header->u.entries, &unpacked_deltas[i]));
+        WT_CELL_FOREACH_DELTA_INT(session, ref->page->dsk, header, unpacked_deltas[i][j])
+        {
+            j++;
+        }
+        WT_CELL_FOREACH_END;
+    }
+
+    WT_ERR(__bt_merge_internal_deltas(session, unpacked_deltas, 0, (uint32_t)delta_size - 1,
+      delta_size_each, &unpacked_deltas_merged, &unpacked_deltas_merged_size));
+
+    WT_ERR(__bt_merge_internal_delta_with_base_image(session, ref, unpacked_deltas_merged,
+      unpacked_deltas_merged_size, &refs, &refs_entries, &incr));
+
+    /*
+     * Constructs a new `p-index` using the merged `refs` list and allocate refs to the new
+     * `p-index`.
+     */
+    pindex_size = sizeof(WT_PAGE_INDEX) + refs_entries * sizeof(WT_REF *);
+    WT_ERR(__wt_calloc(session, 1, pindex_size, &pindex));
+    incr += pindex_size;
+    pindex->index = (WT_REF **)(pindex + 1);
+    pindex->entries = (uint32_t)refs_entries;
+    for (i = 0; i < pindex->entries; ++i) {
+        refs[i]->pindex_hint = i;
+        pindex->index[i] = refs[i];
```

Comment:
Can we just assign refs to pindex->index directly?

Reply:
I find the way we allocate memory for pindex interesting. We always aim to keep the refs in contiguous blocks of pindex. I'm not sure why this approach was originally chosen, but since it's the established method, I want to maintain consistency. Therefore, I allocate the `ref*` array as a contiguous block and then copy the refs over one by one.

Reply:
Makes sense. This helps easing freeing the memory. So you only need to call one free instead of two.